Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -627,4 +627,150 @@ public struct NonBlockingFileIO: Sendable {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            try Posix.unlink(pathname: path)
+        }
+    }
+
+    /// Creates directory at `path` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the directory to be created.
+    ///     - withIntermediateDirectories: Whether intermediate directories should be created.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` which is fulfilled if the rename was successful or fails on error.
+    public func createDirectory(path: String, withIntermediateDirectories createIntermediates: Bool, mode: NIOPOSIXFileMode, eventLoop: EventLoop) -> EventLoopFuture<Void> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            if createIntermediates {
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+                try Posix.mkpath(pathname: path, mode: mode)
+#else
+                // ensure that intermediate directories exist
+                let startIndex = path.startIndex
+                var endIndex = path.index(after: startIndex)
+                while endIndex != path.endIndex {
+                    if path[endIndex] == "/" {
+                        let intermediate = String(path[startIndex..<endIndex])
+                        do {
+                            try Posix.mkdir(pathname: intermediate, mode: S_IRWXU)
+                        } catch let e as IOError where e.errnoCode == EEXIST {
+                            // If intermediate item exists - continue
+                        }
+                    }
+                    endIndex = path.index(after: endIndex)
+                }
+                try Posix.mkdir(pathname: path, mode: mode)
+#endif
+            } else {
+                try Posix.mkdir(pathname: path, mode: mode)
+            }
+        }
+    }
+
+    /// List contents of the directory at `path` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the directory to list the content of.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` containing the file list.
+    public func listDirectory(path: String, eventLoop: EventLoop) -> EventLoopFuture<[String]> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            let dir = try Posix.opendir(pathname: path)
+            var entries: [String] = []
+            do {
+                while let entry = try Posix.readdir(dir: dir) {
+                    let name = withUnsafeBytes(of: entry.pointee.d_name) { pointer -> String in
+                        let ptr = pointer.baseAddress!.assumingMemoryBound(to: CChar.self)
+                        return String(cString: ptr)
```

Comment:
Prefer using buffer pointers and `d_namlen`.

Reply:
Relatedly, should this be slightly richer? Should we include the type info from `dirent`?

Reply:
on linux it's a null-terminated string if I'm not mistaken, for darwin added using a buffer pointer, plus added a struct with 2 additional fields instead of just string